working/4271 - static enough metaprogramming/proposal.md

```diff
@@ -0,0 +1,1222 @@
+some of the errors earlier. However I currently don't see direct analyzer
+support as a necessary requirement to shipping this feature.
+
+> [!NOTE]
+>
+> If we decide that `@konst` is a no-op in debug (development) mode and
+> `@konst` reflection falls back to actual reflection then it is reasonable
+> to expect that developer could still opt-in into full `@konst` evaluation
+> for debug (development) mode. This is important to enable testing this code
+> without forcing developers to build their code in release mode.
+
+> [!NOTE]
+>
+> Some of our tools (most notably analyzer and DDC) are capable of
+> analyzing/compiling libraries only given the _outlines_ of their dependencies.
+> This was one of the showstoppers for enhanced constant proposal, which did not
+> introduce a syntactic marker to delineate code potentially needed for
+> `const`-evaluation from the rest of the program, which would mean that
+> outlines had to contain bodies for most methods - making them impractically
+> large and erasing their benefits.
+>
+> This proposal does not suffer the same issue, even if we decide to support
+> `@konst` evaluation in these tools: as `@konst` actually provides a syntactic
+> marker which partitions the program. Only bodies of methods with `@konst`
+> parameters need to be included into the outline.
+
+### Non-Goals
+
+This proposal does _not_ intend to cover all cases which are supportable via
+Dart source code generation or which are supported by macro systems in other
+programming languages. For example the following capabilities are out of scope:
+
+- injecting new declarations into the program;
+- changing or expanding Dart syntax.
+
+This means that certain things are not possible with this proposal which are
+possible with code generation. Most notably it is not possible to declare
+fields, methods or parameters. This means for example that it is impossible
+to use `@konst` to inject a `copyWith` method based on the list of fields.
+
+> [!NOTE]
+>
+> It is however possible to define method like this:
+>
+> ```dart
+> // Universal function to apply updates specified by the record.
+> T copyWith<@konst T, @konst R extends Record>(T obj, R updates) {
+>   // ...
+> }
+> ```
+>
+> Though usability of this method will be questionable as there will be good
+> autocomplete available for `updates` parameter.
+
+### Prototype implementation
+
+To get the feeling of expressive power, implementation complexity and costs I
+have thrown together a very rough prototype implementation which can be found
+[here][konst-prototype-branch]. When comparing manual toJSON implementation with
+a similar (but not equivalent!) one based on `@konst` reflection I got the
+following numbers:
+
+- Average code size overhead per class: 270 bytes
+- Average JIT kernel generation overhead per-class: 0.2ms (cost of producing
+  specialized functions using Kernel-to-Kernel AST transformation)
+- Average AOT compilation overhead per-class: 2.6ms
```

Comment:
What is the baseline?

Reply:
I don't have numbers right now but I can try to revive the prototype and update this later.